internal/pkg/build/build.go

```diff
@@ -135,14 +139,113 @@ func newBuild(defs []types.Definition, conf Config) (*Build, error) {
 	case "sandbox":
 		b.stages[lastStageIndex].a = &assemblers.SandboxAssembler{}
 	case "sif":
-		b.stages[lastStageIndex].a = &assemblers.SIFAssembler{}
+		mksquashfsPath, err := getMksquashfsPath()
+		if err != nil {
+			return nil, fmt.Errorf("while searching for mksquashfs: %v", err)
+		}
+
+		flag, err := ensureGzipComp(b.stages[lastStageIndex].b.Path, mksquashfsPath)
+		if err != nil {
+			return nil, fmt.Errorf("while ensuring correct compression algorithm: %v", err)
+		}
+		b.stages[lastStageIndex].a = &assemblers.SIFAssembler{
+			GzipFlag:       flag,
+			MksquashfsPath: mksquashfsPath,
+		}
 	default:
 		return nil, fmt.Errorf("unrecognized output format %s", conf.Format)
 	}
 
 	return b, nil
 }
 
+func getMksquashfsPath() (string, error) {
+	// Parse singularity configuration file
+	c := &singularityConfig.FileConfig{}
+	if err := config.Parser(buildcfg.SYSCONFDIR+"/singularity/singularity.conf", c); err != nil {
+		return "", fmt.Errorf("unable to parse singularity.conf file: %s", err)
+	}
+
+	// p is either "" or the string value in the conf file
+	p := c.MksquashfsPath
+
+	// If the path contains the binary name use it as is, otherwise add mksquashfs via filepath.Join
+	if !strings.HasSuffix(c.MksquashfsPath, "mksquashfs") {
+		p = filepath.Join(c.MksquashfsPath, "mksquashfs")
+	}
+
+	// exec.LookPath functions on absolute paths (ignoring $PATH) as well
+	return exec.LookPath(p)
+}
+
+// ensureGzipComp builds dummy squashfs images and checks the type of compression used
+// to deduce if we can successfully build with gzip compression. It returns an error
+// if we cannot and a boolean to indicate if the `-comp` flag is needed to specify
+// gzip compression when the final squashfs is built
+func ensureGzipComp(tmpdir, mksquashfsPath string) (bool, error) {
+	sylog.Debugf("Ensuring gzip compression for mksquashfs")
+
+	var err error
+	s := packer.NewSquashfs()
+	s.MksquashfsPath = mksquashfsPath
+
+	srcf, err := ioutil.TempFile(tmpdir, "squashfs-gzip-comp-test-src")
+	if err != nil {
+		return false, fmt.Errorf("while creating temporary file for squashfs source: %v", err)
+	}
+
+	srcf.Write([]byte("Test File Content"))
+	srcf.Close()
+
+	f, err := ioutil.TempFile(tmpdir, "squashfs-gzip-comp-test-")
+	if err != nil {
+		return false, fmt.Errorf("while creating temporary file for squashfs: %v", err)
```

Comment:
this is leaving the temporary file from line 192 behind.

Reply:
Both of these files will be taken care of when the build cleans up overall. I can see a potential debugging case where we would want at least the test squashfs to stay around if we use `--no-cleanup` in case we have a squashfs tool related error

Reply:
ah, I missed the tmpdir argument, thanks